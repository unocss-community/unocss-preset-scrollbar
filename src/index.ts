import type { Preset } from 'unocss'
import { colorResolver, handler } from '@unocss/preset-mini/utils'

const defaultOption: Required<PresetScrollbarDefaultOption> = {
  scrollbarWidth: '8px',
  scrollbarHeight: '8px',
  scrollbarTrackRadius: '4px',
  scrollbarThumbRadius: '4px',
  scrollbarTrackColor: '#f5f5f5',
  scrollbarThumbColor: '#ddd',
  varPrefix: '',
  numberToUnit: value => `${value / 4}rem`,
}

export interface PresetScrollbarDefaultOption {
  /**
   * default scrollbar width
   * @default '8px'
   */
  scrollbarWidth?: string
  /**
   * default scrollbar height
   * @default '8px'
   */
  scrollbarHeight?: string
  /**
   * default scrollbar track radius
   * @default '4px'
   */
  scrollbarTrackRadius?: string
  /**
   * default scrollbar thumb radius
   * @default '4px'
   */
  scrollbarThumbRadius?: string
  /**
   * default scrollbar track background color
   * @default '#f5f5f5'
   */
  scrollbarTrackColor?: string
  /**
   * default scrollbar thumb background color
   * @default '#ddd'
   */
  scrollbarThumbColor?: string
  /**
   * css variable prefix
   * @default ''
   */
  varPrefix?: string
  /**
   * convert number to unit
   * @default value => `${value / 4}rem`
   * @example
   * numberToUnit: value => `${value / 4}rem`
   * p-4 => padding: 1rem
   * p-4px => padding: 4px
   *
   * @example
   * numberToUnit: value => `${value}rpx`
   * p-4 => padding: 4rpx
   */
  numberToUnit?: (value: number) => string
}

const customRules = {
  'radius': ['track-radius', 'thumb-radius'],
  'w': ['width'],
  'h': ['height'],
  'track-radius': ['track-radius'],
  'thumb-radius': ['thumb-radius'],
}

const numberVarRegex = new RegExp(`^scrollbar-(${Object.keys(customRules).join('|')})-(\\d+?)([a-zA-Z]*?)$`)

export function presetScrollbar(option?: PresetScrollbarDefaultOption): Preset {
  const config = {
    ...defaultOption,
    ...option,
  }

  function resolveVar(name: string) {
    const prefix = config.varPrefix
    return `--${prefix ? `${prefix}-` : ''}scrollbar-${name}`
  }

  const variantsRE = /^(scrollbar(-track|-thumb)?):.+$/

  return {
    name: 'unocss-preset-scrollbar',
    shortcuts: [
      [
        'scrollbar', [
          { overflow: 'auto' },
          'scrollbar-custom-property',
          `scrollbar-color-[var(${resolveVar('thumb')})_var(${resolveVar('track')})]`,
          `scrollbar-track:scrollbar-background-color-[var(${resolveVar('track')})]`,
          `scrollbar-thumb:scrollbar-background-color-[var(${resolveVar('thumb')})]`,
          `scrollbar:scrollbar-width-[var(${resolveVar('width')})]`,
          `scrollbar:scrollbar-height-[var(${resolveVar('height')})]`,
        ],
      ],
      [
        'scrollbar-rounded', `
          scrollbar-track:scrollbar-border-radius-[var(${resolveVar('track-radius')})]
          scrollbar-thumb:scrollbar-border-radius-[var(${resolveVar('thumb-radius')})]
        `,
      ],
    ],
    variants: [
      // ::-webkit-scrollbar-track
      // ::-webkit-scrollbar-thumb
      // ::-webkit-scrollbar
      (matcher) => {
        if (!variantsRE.test(matcher))
          return

        const variant = matcher.replace(variantsRE, '$1')

        return {
          matcher: matcher.slice(variant.length + 1),
          selector: (s) => {
            return `${s}::-webkit-${variant}`
          },
        }
      },
    ],
    rules: [
      [
        /^scrollbar-color-(.+)$/,
        ([_, prop]) => ({
          'scrollbar-color': handler.bracket.cssvar.auto.fraction.rem(prop),
        }),
      ],
      [
        /^scrollbar-custom-property$/,
        ([_]) => ({
          [resolveVar('track')]: config.scrollbarTrackColor,
          [resolveVar('thumb')]: config.scrollbarThumbColor,
          [resolveVar('width')]: config.scrollbarWidth,
          [resolveVar('height')]: config.scrollbarHeight,
          [resolveVar('track-radius')]: config.scrollbarTrackRadius,
          [resolveVar('thumb-radius')]: config.scrollbarThumbRadius,
        }),
      ],
      [
        /^scrollbar-thumb-color-(.+)$/,
        colorResolver(resolveVar('thumb'), 'scrollbar-thumb'),
        { autocomplete: 'scrollbar-thumb-color-$colors' },
      ],
      [
        /^scrollbar-track-color-(.+)$/,
        colorResolver(resolveVar('track'), 'scrollbar-track'),
        { autocomplete: 'scrollbar-track-color-$colors' },
      ],
      [
        /^scrollbar-(width|height|background\-color|border\-radius|)-(\[var.+\])$/,
        ([_, prop, value]) => {
          return {
            [`${prop}`]: handler.bracket(value),
          }
        },
      ],
      [
        numberVarRegex,
        ([_, type, value, unit]) => {
          const val = unit ? value + unit : config.numberToUnit(Number.parseInt(value))
          const vars = customRules[type as keyof typeof customRules]
            .map(v => resolveVar(v))
          return vars.reduce((acc: any, v) => {
            acc[v] = val
            return acc
          }, {})
        },
        { autocomplete: `scrollbar-(${Object.keys(customRules).join('|')})-<num>` },
      ],
    ],
  }
}
