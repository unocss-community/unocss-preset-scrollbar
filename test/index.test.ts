import { createGenerator, presetAttributify, presetUno } from 'unocss'
import { describe, expect, it } from 'vitest'
import { createAutocomplete } from '@unocss/autocomplete'
import { presetScrollbar } from '../src'

describe('scrollbar', () => {
  const generator = createGenerator({
    presets: [
      presetUno({
        preflight: false,
      }),
      presetAttributify(),
      presetScrollbar(),
    ],
  })

  const ac = createAutocomplete(generator)

  async function enumerateSuggestions(inputs: string[]) {
    return Object.fromEntries(await Promise.all(inputs.map(async input => [
      input,
      (await ac.suggest(input)).slice(0, 10).join(' '),
    ])))
  }

  it('scrollbar', async () => {
    const { css } = await generator.generate([
      'scrollbar',
      'scrollbar-rounded',
      'scrollbar-w-4px',
      'scrollbar-radius-2',
      'scrollbar-radius-track-4',
      'scrollbar-radius-thumb-4',
    ].join(' '))

    expect(css).toMatchSnapshot()
  })

  it('scrollbar color', async () => {
    const { css } = await generator.generate([
      'scrollbar-track-color-red',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('scrollbar custom unit', async () => {
    const { css } = await generator.generate([
      'scrollbar-w-1px',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('custom value to unit', async () => {
    const generator = createGenerator({
      presets: [
        presetUno({
          preflight: false,
        }),
        presetScrollbar({
          numberToUnit: value => `${value * 2}px`,
        }),
      ],
    })
    const { css } = await generator.generate([
      'scrollbar-w-1',
      'scrollbar-thumb-radius-2px',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('should work in atributify mode', async () => {
    const { css } = await generator.generate(`
<div 
  scrollbar="~ rounded w-4px radius-2 radius-track-4 radius-thumb-4">
</div>
`)
    expect(css).toMatchSnapshot()
  })

  it('var prefix', async () => {
    const generator = createGenerator({
      presets: [
        presetUno({
          preflight: false,
        }),
        presetScrollbar({
          varPrefix: 'my-custom-prefix',
        }),
      ],
    })
    const { css } = await generator.generate([
      'scrollbar',
      'scrollbar-w-1',
      'scrollbar-thumb-radius-2px',
      'scrollbar-rounded',
    ].join(' '))
    expect(css).toMatchSnapshot()
  })

  it('should provide autocomplete', async () => {
    expect(
      await enumerateSuggestions([
        'scrollbar-',
        'scrollbar-w-',
        'sccrollbar-thumb-radius-',
        'scrollbar-radius-',
        'scrollbar-track-color-',
        'scrollbar-thumb-color-',
        'scrollbar-thumb-',
      ]),
    ).toMatchInlineSnapshot(`
      {
        "sccrollbar-thumb-radius-": "",
        "scrollbar-": "scrollbar-rounded scrollbar-track-color-red scrollbar-w-4px scrollbar-radius-2 scrollbar-w-1px",
        "scrollbar-radius-": "scrollbar-radius-2 scrollbar-radius-0 scrollbar-radius-1 scrollbar-radius-3 scrollbar-radius-4 scrollbar-radius-5 scrollbar-radius-6 scrollbar-radius-8 scrollbar-radius-10 scrollbar-radius-12",
        "scrollbar-thumb-": "",
        "scrollbar-thumb-color-": "scrollbar-thumb-color-amber scrollbar-thumb-color-black scrollbar-thumb-color-blue scrollbar-thumb-color-bluegray scrollbar-thumb-color-blueGray scrollbar-thumb-color-coolgray scrollbar-thumb-color-coolGray scrollbar-thumb-color-current scrollbar-thumb-color-cyan scrollbar-thumb-color-dark",
        "scrollbar-track-color-": "scrollbar-track-color-amber scrollbar-track-color-black scrollbar-track-color-blue scrollbar-track-color-bluegray scrollbar-track-color-blueGray scrollbar-track-color-coolgray scrollbar-track-color-coolGray scrollbar-track-color-current scrollbar-track-color-cyan scrollbar-track-color-dark",
        "scrollbar-w-": "scrollbar-w-4px scrollbar-w-1px scrollbar-w-0 scrollbar-w-1 scrollbar-w-2 scrollbar-w-3 scrollbar-w-4 scrollbar-w-5 scrollbar-w-6 scrollbar-w-8",
      }
    `)
  })

  it('preset prefix', async () => {
    const generator = createGenerator({
      presets: [
        presetUno({
          preflight: false,
          prefix: 'tw-',
        }),
        presetScrollbar(),
      ],
    }) 
    const {
      css,
    } = await generator.generate([
      'scrollbar',
    ])
    expect(css).toMatchSnapshot()
  })
})
