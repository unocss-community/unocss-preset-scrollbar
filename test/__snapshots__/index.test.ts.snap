// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scrollbar > custom value to unit 1`] = `
"/* layer: default */
.scrollbar-thumb-radius-2px{--scrollbar-thumb-radius:2px;}
.scrollbar-w-1{--scrollbar-width:2px;}"
`;

exports[`scrollbar > preset set prefix 1`] = `
"/* layer: default */
.un-scrollbar{--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
.un-scrollbar::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
.un-scrollbar::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
.un-scrollbar::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}"
`;

exports[`scrollbar > presetUno prefix 1`] = `
"/* layer: shortcuts */
.scrollbar{--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
.scrollbar::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
.scrollbar::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
.scrollbar::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}"
`;

exports[`scrollbar > scrollbar 1`] = `
"/* layer: shortcuts */
.scrollbar{--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
.scrollbar::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
.scrollbar::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
.scrollbar::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}
.scrollbar-rounded::-webkit-scrollbar-thumb{border-radius:var(--scrollbar-thumb-radius);}
.scrollbar-rounded::-webkit-scrollbar-track{border-radius:var(--scrollbar-track-radius);}
/* layer: default */
.scrollbar-radius-2{--scrollbar-track-radius:0.5rem;--scrollbar-thumb-radius:0.5rem;}
.scrollbar-w-4px{--scrollbar-width:4px;}"
`;

exports[`scrollbar > scrollbar color 1`] = `
"/* layer: default */
.scrollbar-track-color-red{--un-scrollbar-track-opacity:1;--scrollbar-track:rgb(248 113 113 / var(--un-scrollbar-track-opacity));}"
`;

exports[`scrollbar > scrollbar custom unit 1`] = `
"/* layer: default */
.scrollbar-w-1px{--scrollbar-width:1px;}"
`;

exports[`scrollbar > should work in atributify mode 1`] = `
"/* layer: shortcuts */
[scrollbar~="\\~"]{--scrollbar-track:#f5f5f5;--scrollbar-thumb:#ddd;--scrollbar-width:8px;--scrollbar-height:8px;--scrollbar-track-radius:4px;--scrollbar-thumb-radius:4px;overflow:auto;}
[scrollbar~="\\~"]::-webkit-scrollbar{width:var(--scrollbar-width);height:var(--scrollbar-height);}
[scrollbar~="\\~"]::-webkit-scrollbar-thumb{background-color:var(--scrollbar-thumb);}
[scrollbar~="\\~"]::-webkit-scrollbar-track{background-color:var(--scrollbar-track);}
[scrollbar~="rounded"]::-webkit-scrollbar-thumb{border-radius:var(--scrollbar-thumb-radius);}
[scrollbar~="rounded"]::-webkit-scrollbar-track{border-radius:var(--scrollbar-track-radius);}
/* layer: default */
.w-4px{width:4px;}
.rounded{border-radius:0.25rem;}
[scrollbar~="radius-2"]{--scrollbar-track-radius:0.5rem;--scrollbar-thumb-radius:0.5rem;}
[scrollbar~="w-4px"]{--scrollbar-width:4px;}"
`;

exports[`scrollbar > var prefix 1`] = `
"/* layer: shortcuts */
.scrollbar{--my-custom-prefix-scrollbar-track:#f5f5f5;--my-custom-prefix-scrollbar-thumb:#ddd;--my-custom-prefix-scrollbar-width:8px;--my-custom-prefix-scrollbar-height:8px;--my-custom-prefix-scrollbar-track-radius:4px;--my-custom-prefix-scrollbar-thumb-radius:4px;overflow:auto;}
.scrollbar::-webkit-scrollbar{width:var(--my-custom-prefix-scrollbar-width);height:var(--my-custom-prefix-scrollbar-height);}
.scrollbar::-webkit-scrollbar-thumb{background-color:var(--my-custom-prefix-scrollbar-thumb);}
.scrollbar::-webkit-scrollbar-track{background-color:var(--my-custom-prefix-scrollbar-track);}
.scrollbar-rounded::-webkit-scrollbar-thumb{border-radius:var(--my-custom-prefix-scrollbar-thumb-radius);}
.scrollbar-rounded::-webkit-scrollbar-track{border-radius:var(--my-custom-prefix-scrollbar-track-radius);}
/* layer: default */
.scrollbar-thumb-radius-2px{--my-custom-prefix-scrollbar-thumb-radius:2px;}
.scrollbar-w-1{--my-custom-prefix-scrollbar-width:0.25rem;}"
`;
